File: schemas/social-media/convex/follow_requests.ts
// follow_requests: pending follow requests for private profiles with approval workflow.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const followRequests = defineTable({
  requesterId: v.id("users"),
  targetId: v.id("users"),
  status: v.union(v.literal("pending"), v.literal("approved"), v.literal("rejected")),
  updatedAt: v.number(),
})
  .index("by_requester_id_target_id", ["requesterId", "targetId"])
  .index("by_target_id_status", ["targetId", "status"]);

File: schemas/social-media/spacetimedb/follow_requests.rs
// follow_requests: pending, approved, or rejected follow requests for private profiles.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum FollowRequestStatus {
    Pending, // type: String
    Approved,
    Rejected,
}

#[spacetimedb::table(name = follow_requests, public)]
pub struct FollowRequest {
    #[primary_key]
    pub id: String, // UUID
    // Composite unique: (requester_id, target_id) — enforce in reducer logic.
    #[index(btree)]
    pub requester_id: String, // UUID — FK → users.id (cascade delete)
    // Composite index: (target_id, status) — not supported in SpacetimeDB; document only.
    #[index(btree)]
    pub target_id: String, // UUID — FK → users.id (cascade delete)
    pub status: FollowRequestStatus,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
